### PR TITLE
Deprecation warning in _replace_vars_and_dims

### PR DESCRIPTION
**Pull request by max-sixty**
& removed internal usages

 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

I needed to pass `indexes=None` from some of the existing methods that were using the deprecated method. I spent some time debugging but couldn't work out the causality. I added TODOs as part of the index work.

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -962,6 +962,11 @@ def _replace_vars_and_dims(
         Unlike _replace_with_new_dims(), this method always recalculates
         indexes from variables.
         """
+        warnings.warn(
+            "`_replace_vars_and_dims` is deprecated in favor of `_replace_with_new_dims` and will be removed in 0.16",
+            DeprecationWarning,
+            stacklevel=2,
+        )
```
Comment (shoyer) [inline xarray/core/dataset.py]: Do we really need to deprecate this? It's an internal API, so I think we should feel free to remove it without warning.

Comment (max-sixty) [reply xarray/core/dataset.py]: Yeah, I wasn't sure, and thought it safer to deprecate, but I was probably being overly cautious. I'll remove.

**Comment by max-sixty** [comment]
Updated!

**Comment by max-sixty** [comment]
(test failure unrelated)

File: xarray/backends/zarr.py [hunk 2]
```diff
@@ -656,4 +656,4 @@ def maybe_chunk(name, var, chunks):
             return var
 
     variables = {k: maybe_chunk(k, v, chunks) for k, v in ds.variables.items()}
-    return ds._replace_vars_and_dims(variables)
+    return ds._replace_with_new_dims(variables)
```
Comment (dcherian) [inline xarray/backends/zarr.py]: can this just be `_replace`? I find it confusing to "replace with new dims" but not specify "new dims".

File: xarray/core/dataset.py [hunk 3]
```diff
@@ -941,25 +941,6 @@ def _replace_with_new_dims(
...
-    ) -> "Dataset":
-        """Deprecated version of _replace_with_new_dims().
-
-        Unlike _replace_with_new_dims(), this method always recalculates
-        indexes from variables.
-        """
-        if dims is None:
-            dims = calculate_dimensions(variables)
-        return self._replace(
-            variables, coord_names, dims, attrs, indexes=None, inplace=inplace
```
Comment (dcherian) [inline xarray/core/dataset.py]: looks like this old function set `indexes=None` so your later changes make sense.

Also see docstring: 
`Unlike _replace_with_new_dims(), this method always recalculates indexes from variables.`

File: xarray/backends/zarr.py (hunk 2 above)
Comment (max-sixty) [reply xarray/backends/zarr.py]: Hmmm, I think in this case `dims` need to be recomputed, since `variables` changed?

**Comment by shoyer** [comment]
Looking more carefully at this, I'm remembering now why I called this "deprecated." Eventually, we always want to keep track of `indexes` explicitly. Currently we don't do that, and in particular `_replace_vars_and_dims` does not do that.

My main reason for keeping it around was simply as a reminder that the functions that use it still need to be updated. It would be great to do that, but each usage should explicitly set `indexes` rather than setting `indexes=None` which we want to eventually stop doing.

Does that make sense?

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -3024,11 +3005,11 @@ def expand_dims(
...
-        new_dims = self._dims.copy()
-        new_dims.update(dim)
-
-        return self._replace_vars_and_dims(
-            variables, dims=new_dims, coord_names=coord_names
+        return self._replace_with_new_dims(
+            # TODO: fix indexes, I think by passing nothing (i.e. default)
+            variables=variables,
+            coord_names=coord_names,
+            indexes=None,
```
Comment (shoyer) [inline xarray/core/dataset.py]: I think the right way to set `indexes` here would be to copy them from the original dataset and then add/modify indexes as needed for any expanded dimensions.

**Pull request closed without merging**